### PR TITLE
✨ Feature ➾ Improve SmartPy Workflow in relation to contract and storage compilations and added native SmartPy test

### LINKED ISSUE #1475: 🚧 Dev ➾ Bring SmartPy Plugin up to Par with Ligo Plugin Functionality
### 🔗 Epic|Feature Link

https://ecadlabs.productboard.com/feature-board/5219572-taqueria-features/features/14717143/detail

### 🚥 Definition of Done

- [X] 🎯 Planning and Design
- [ ] 🏎️ Dev Implementation Complete
- [ ] ♾️ PR Merged
- [ ] 🏆 Product Acceptance

### 🧰 Work Description

In support of the Milestone 3 deliverable for SmartPy users can continuously develop apps, this ticket takes the first step towards an improved SmartPy Developer workflow by bringing the functionality up to par with the Ligo plugin

### ✅ To Do

- [x] Abstraction over contract and storage compilations.
- [x] Add expression compilation (potentially for parameters).
- [x] Native test.
- [x] Doc updates.
- [ ] Make it easier to install smartpy
- [ ] Better version management of smartpy. Maybe they or we can dockerize it.
- [ ] Arbitrary command exec.
- [ ] Better error messaging.

### ⚖️ Acceptance Criteria

✨ Functional: 
- [x] All tasks provided by the Ligo plugin are available in SmartPy
- [ ] The version of SmartPy is updated
- [x] Native tests are supported
- [x] Storage and parameter definition is brought in line with the naming schema
- [ ] Investigation of moving plugin to Docker

### 🔭 Scope

🙅‍♂️ Out of Scope: 
- Work on Ligo plugin
 
 

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by jchenche**
# 🌮 Taqueria PR

Fixes #1475 

## 🪁 Description

Provide a smooth SmartPy workflow for our SmartPy plugin by making it easier to compile and deploy. E.g. `taq compile my_contract.py` then `taq deploy my_contract.tz` (or `taq deploy my_contract.tz --env testing` to deploy to ghostnet) and that's it.

Users will define all compilation and expression compilation targets in their main SmartPy contract file and we will compile them all and copy the relevant artifacts (contract, storage, and expression files) to the `/artifacts` folder. All other artifacts will be hidden under `/artifacts/.smartpy`.

Compilation targets in SmartPy produce both the contract and its storage Michelson files. We will name them using the following naming convention: `<CONTRACT_NAME>.storage.<COMPILATION_TARGET_NAME>.tz`. But the first of these is special and will be called: `<CONTRACT_NAME>.default_storage.tz` (the Taquito deploy task will deem this as the default initial storage value).

Expression compilation targets in SmartPy produce only the expression Michelson file. We will name them using the following naming convention: `<CONTRACT_NAME>.expression.<COMPILATION_TARGET_NAME>.tz`.

Native SmartPy testing can now be done with `taq test my_contract.py`.

In SmartPy and LIGO's compile task, `--json` can be specified to emit JSON-encoded Michelson instead of pure Michelson expressions.

## 🪂 Pre-Merge Checklist (Definition of Done)

🚦 Required to merge:

- [x] ⛱️ I have completed this PR template in full and updated the title appropriately
- [x] ⛵ My code builds cleanly, and I have manually tested the changes
- [ ] 🏄‍♂️ Another team member has built this branch and done manual testing on the change
- [x] 🏖️ New and existing unit tests pass locally and in CI
- [ ] 🦀 Automated tests have been written and added to this PR
- [x] 🐬 I have commented my code, particularly in hard-to-understand areas
- [x] 🤿 Corresponding changes have been made to all documentation
- [x] 🚢 The release checklist has been completed

## 🛩️ Summary of Changes

_Please include a summary of the changes to the codebase._ _Please also include relevant motivation and context for the change_
_(e.g. what was the existing behaviour, why did it break, etc.)_

## 🎢 Test Plan

_Please describe the testing strategy and plan for this PR. Keep this lightweight and anticipate any testing challenges_

## 🛸 Type of Change

- [ ] 🧽 Chore ➾
- [ ] 🛠️ Fix ➾
- [x] ✨ Feature ➾
- [x] 👷 Refactor ➾
- [ ] 🧪 Pre-Release ➾
- [ ] 🚀 Release ➾

### CODE REVIEW

File: taqueria-plugin-smartpy/compile.ts [hunk 1]
```diff
@@ -1,66 +1,185 @@
...
+type TableRow = { contract: string; artifact: string };
+
+const COMPILE_ERR_MSG: string = 'Not compiled';
+
+const addPyExtensionIfMissing = (sourceFile: string): string => {
+	return /\.py$/.test(sourceFile) ? sourceFile : `${sourceFile}.py`;
+};
+
+const getOutputContractFilename = (parsedArgs: Opts, sourceFile: string): string => {
+	const outputFile = basename(sourceFile, extname(sourceFile));
```
Comment (AlirezaHaghshenas) [inline taqueria-plugin-smartpy/compile.ts:15]: I think the function `basename` will just do the same if you don't pass the second argument to it? And will be simpler.

Comment (jchenche) [reply taqueria-plugin-smartpy/compile.ts:17]: It won't do the same. The second argument is needed

**Review by AlirezaHaghshenas — APPROVED** [review]

File: taqueria-plugin-smartpy/compile.ts (hunk 1 above)
Comment (AlirezaHaghshenas) [reply taqueria-plugin-smartpy/compile.ts:17]:
```ts
            /**
             * Return the last portion of a path. Similar to the Unix basename command.
             * Often used to extract the file name from a fully qualified path.
             *
             * @param path the path to evaluate.
             * @param suffix optionally, an extension to remove from the result.
             * @throws {TypeError} if `path` is not a string or if `ext` is given and is not a string.
             */
            basename(path: string, suffix?: string): string;
```

Comment (jchenche) [reply taqueria-plugin-smartpy/compile.ts:17]: Yep there `@param suffix optionally, an extension to remove from the result.`

path1/path2/file.mligo becomes file.mligo but with that 2nd argument, it becomes file, which is what I want

**Review by AlirezaHaghshenas — APPROVED** [review]

**Review by AlirezaHaghshenas — APPROVED** [review]

File: taqueria-plugin-smartpy/common.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import { sendErr } from '@taqueria/node-sdk';
```
Comment (AlirezaHaghshenas) [inline taqueria-plugin-smartpy/common.ts:1]: Some functions in this file are very similar to the one in ligo plugin. Is there an easy way to unify them?

Comment (jchenche) [reply taqueria-plugin-smartpy/common.ts:1]: I can put them in the SDK but they are only used for compilers. Doesn't seem necessary for this amount of code.

**Review by AlirezaHaghshenas — APPROVED** [review]

**Review by mweichert — CHANGES_REQUESTED** [review]
There was just one missing `await` statement in the test suite. Other than that, looks good.

Side note, although it doesn't impact the test results when they run correctly, wrapping your test in a try/catch block is unnecessary and actually makes it harder to see where the failure occurred , as the line number of what assertion isn't present.

E.g.:

```
test('Verify that taqueria SmartPy plugin will output proper error message running taq test <sourceFile> command against invalid test file', async () => {
		try {
			// 1. Copy contract from data folder to taqueria project folder
			await exec(`cp e2e/data/hello-tacos-failed-tests.py ${taqueriaProjectPath}/contracts`);

			// 2. Run taq test ${testFileName}
			const { stdout } = await exec(`taq test hello-tacos-failed-tests.py`, {
				cwd: `./${taqueriaProjectPath}`,
			});
			expect(stdout).toContain('Some tests failed');
		} catch (error) {
			throw new Error(`error: ${error}`);
		}
	});
```

Should be:

```
test('Verify that taqueria SmartPy plugin will output proper error message running taq test <sourceFile> command against invalid test file', async () => {
	// 1. Copy contract from data folder to taqueria project folder
	await exec(`cp e2e/data/hello-tacos-failed-tests.py ${taqueriaProjectPath}/contracts`);

	// 2. Run taq test ${testFileName}
	const { stdout } = await exec(`taq test hello-tacos-failed-tests.py`, {
		cwd: `./${taqueriaProjectPath}`,
	});
	expect(stdout).toContain('Some tests failed');
});
```

This is something I had talked to Alex, and I've been removing them as I come across them. But what I think would be better is to create an issue that involves going through each test suite and removing those try/catch wrappers.

File: tests/e2e/taqueria-plugin-smartpy.spec.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		} catch (error) {
+			throw new Error(`error: ${error}`);
+		}
+	});
+
+	// Clean up process to remove taquified project folder
+	// Comment if need to debug
+	afterAll(async () => {
+		try {
+			fsPromises.rm(taqueriaProjectPath, { recursive: true });
```
[the remark this answers is not in the document]
Comment (jchenche) [reply tests/e2e/taqueria-plugin-smartpy.spec.ts:36]: got it!

**Review by hu3man — APPROVED** [review]

**Pull request merged**